PG-1419: Support processing of USFM3 quote start/end milestones

tombogle opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/sillsdev/glyssen/763)
<!-- Reviewable:end -->

andrew-polk reviewed (changes requested):
I've only done the easy stuff so far, but my brain is dead on a Friday pm...

> Reviewed 29 of 33 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/glyssen/763#-:-N5w1SCwBo9BdgKmPmpm:b9f1w0l)* status: 29 of 33 files reviewed, 8 unresolved discussions (waiting on @tombogle)

___
*[`Glyssen.sln.DotSettings` line 64 at r1](https://reviewable.io/reviews/sillsdev/glyssen/763#-N5q4WCGB53WCEluxdem:-N5q4WCGB53WCEluxden:b-t7yf2p) ([raw file](https://github.com/sillsdev/glyssen/blob/dfd9f63cca8aa193ff73601f3d08f85efa399be7/Glyssen.sln.DotSettings#L64)):*
> ```
> 	<s:Boolean x:Key="/Default/UserDictionary/Words/=versioning/@EntryIndexedValue">True</s:Boolean>
> 	<s:Boolean x:Key="/Default/UserDictionary/Words/=XLIFF/@EntryIndexedValue">True</s:Boolean>
> 	<s:Boolean x:Key="/Default/UserDictionary/Words/=_0009_0009va/@EntryIndexedValue">True</s:Boolean>
> ```

Guessing you don't actually want to add this to the dictionary. Not sure about "rstl" either.



_Code quote:_
```
_0009_0009va
```
___
*[`GlyssenEngine/Project.cs` line 522 at r1](https://reviewable.io/reviews/sillsdev/glyssen/763#-N5vmyMFEOmGMRrlyI1D:-N5vmyMFEOmGMRrlyI1E:bj0lzzh) ([raw file](https://github.com/sillsdev/glyssen/blob/dfd9f63cca8aa193ff73601f3d08f85efa399be7/GlyssenEngine/Project.cs#L522)):*
> ```Smalltalk
> 		}
> 
> 		public string GetStandardCharacterName(string character, string bookId, int chapter,
> ```

Is this supposed to be here?
___
*[`GlyssenEngine/Character/CharacterUsageStore.cs` line 14 at r1](https://reviewable.io/reviews/sillsdev/glyssen/763#-N5vtyL7AOd2b1SxhNmU:-N5vtyL7AOd2b1SxhNmV:b-5a6g9q) ([raw file](https://github.com/sillsdev/glyssen/blob/dfd9f63cca8aa193ff73601f3d08f85efa399be7/GlyssenEngine/Character/CharacterUsageStore.cs#L14)):*
> ```Smalltalk
> namespace GlyssenEngine.Character
> {
> 	internal class CharacterUsageStore : ICharacterUsageStore
> ```

This class would benefit from a class comment.

I'm unsure about the name. But I definitely don't have anything better at this point. Maybe the class comment will help...
___
*[`GlyssenEngine/Character/CharacterUsageStore.cs` line 30 at r1](https://reviewable.io/reviews/sillsdev/glyssen/763#-N5vwKgDCklM_t8zdZ2S:-N5vwKgDCklM_t8zdZ2T:b-vpm1n1) ([raw file](https://github.com/sillsdev/glyssen/blob/dfd9f63cca8aa193ff73601f3d08f85efa399be7/GlyssenEngine/Character/CharacterUsageStore.cs#L30)):*
> ```Smalltalk
> 		}
> 
> 		private class FuzzyMatchCandidate
> ```

This class could also probably benefit from a comment.
___
*[`GlyssenEngine/Character/CharacterUsageStore.cs` line 90 at r1](https://reviewable.io/reviews/sillsdev/glyssen/763#-N5vvrBZ31lOo68MCaY0:-N5vvrBZ31lOo68MCaY1:buse9dt) ([raw file](https://github.com/sillsdev/glyssen/blob/dfd9f63cca8aa193ff73601f3d08f85efa399be7/GlyssenEngine/Character/CharacterUsageStore.cs#L90)):*
> ```Smalltalk
> 			}
> 
> 			// Finally, try for a fuzzy match. The following "magic numbers" are based on some
> ```

you left me hanging...

Actually looks like this second sentence just needs to be deleted. You have it again (and complete) later.



_Code quote:_
```Smalltalk
The following "magic numbers" are based on some
```
___
*[`GlyssenEngine/Script/Block.cs` line 899 at r1](https://reviewable.io/reviews/sillsdev/glyssen/763#-N5vsC-xC2_DbMHZGvnq:-N5vsC-y3BEY7Cui0IDK:b-1b5fa6) ([raw file](https://github.com/sillsdev/glyssen/blob/dfd9f63cca8aa193ff73601f3d08f85efa399be7/GlyssenEngine/Script/Block.cs#L899)):*
> ```Smalltalk
> 		public bool IsQuote => !CharacterVerseData.IsCharacterStandard(CharacterId) ||
> 			UserConfirmed ||
> 			// REVIEW: If this last condition is what makes this a quote, should we also ensure
> ```

Seems like that is the purpose, yes (to indicate a quote as opposed to narrator).



_Code quote:_
```Smalltalk
			// REVIEW: If this last condition is what makes this a quote, should we also ensure
			// that the character is not narrator?
```
___
*[`GlyssenEngineTests/Character/CharacterUsageStoreTests.cs` line 224 at r1](https://reviewable.io/reviews/sillsdev/glyssen/763#-N5w-Ma72N69BLf3imQq:-N5w-Ma72N69BLf3imQr:bvslhqx) ([raw file](https://github.com/sillsdev/glyssen/blob/dfd9f63cca8aa193ff73601f3d08f85efa399be7/GlyssenEngineTests/Character/CharacterUsageStoreTests.cs#L224)):*
> ```Smalltalk
> 		[TestCase("REV", 1, "17", "Son of man", ExpectedResult = "Jesus")]
> 		[TestCase("ISA", 45, "24", "tongue", ExpectedResult = "every tongue")]
> 		public string GetStandardCharacterName_OnlyMatchingSubstring_ReturnsStandardCharacterName(
> ```

My pea-brain needs help on this one. In what sense are the first 3 cases "OnlyMatchingSubstring"?
Those are substrings of an alias perhaps?

Generally speaking for all these tests... is there something which can be done to verify we have the characterID/alias we expect in the test control file?
___
*[`GlyssenShared/BlockElement.cs` line 98 at r1](https://reviewable.io/reviews/sillsdev/glyssen/763#-N5vlm1vCL3Ade-eiYaR:-N5vlm1vCL3Ade-eiYaS:b-ps38n6) ([raw file](https://github.com/sillsdev/glyssen/blob/dfd9f63cca8aa193ff73601f3d08f85efa399be7/GlyssenShared/BlockElement.cs#L98)):*
> ```Smalltalk
> 		public bool ContainsNoWords => !Content.Any(IsLetterOrDigit);
> 
> 		// In almost all situations, this is the opposite of ContainsNoWords, except perhaps for
> ```

Don't you mean the same as ContainsNoWords?



_Code quote:_
```Smalltalk
this is the opposite of ContainsNoWords
```


<!-- Sent from Reviewable.io -->

tombogle commented:
*[`Glyssen.sln.DotSettings` line 64 at r1](https://reviewable.io/reviews/sillsdev/glyssen/763#-N5q4WCGB53WCEluxdem:-N5wXYoh6z92ikcU5M4N:bg4ul65) ([raw file](https://github.com/sillsdev/glyssen/blob/dfd9f63cca8aa193ff73601f3d08f85efa399be7/Glyssen.sln.DotSettings#L64)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Guessing you don't actually want to add this to the dictionary. Not sure about "rstl" either.
</blockquote></details>

Yeah, I've noticed that it sometimes adds nonsensical entries. I caught it in the act once, but it's easy to overlook.


<!-- Sent from Reviewable.io -->

tombogle commented:
*[`GlyssenEngine/Project.cs` line 522 at r1](https://reviewable.io/reviews/sillsdev/glyssen/763#-N5vmyMFEOmGMRrlyI1D:-N5wZmPL1FWD6gF8wX1I:b32w12) ([raw file](https://github.com/sillsdev/glyssen/blob/dfd9f63cca8aa193ff73601f3d08f85efa399be7/GlyssenEngine/Project.cs#L522)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Is this supposed to be here?
</blockquote></details>

No



_Code quote:_
```Smalltalk
public string GetStandardCharacterName(string character, string bookId, int chapter,
```


<!-- Sent from Reviewable.io -->

tombogle commented:
*[`GlyssenEngine/Script/Block.cs` line 899 at r1](https://reviewable.io/reviews/sillsdev/glyssen/763#-N5vsC-xC2_DbMHZGvnq:-N5zw6nf6-WRr94v1Bgd:b-mc8olg) ([raw file](https://github.com/sillsdev/glyssen/blob/dfd9f63cca8aa193ff73601f3d08f85efa399be7/GlyssenEngine/Script/Block.cs#L899)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Seems like that is the purpose, yes (to indicate a quote as opposed to narrator).
</blockquote></details>

Looking more carefully though, we already return true for any non-narrator, so that wouldn't add anything. I guess the question is whether there is some situation where the character ID is not yet set correctly but the style tag tells us we're in a quote, but it's actually an explicit narrator "quote". This code is already (as described) not guaranteed to be 100% reliable. It turns out that the final condition is not needed for any unit tests. So I have commented it out and changed the REVIEW comment to ask whether it is needed at all. I might have just added it for completeness' sake, but it's probably more likely that I added it at some point when it seemed like it would actually matter, but then it didn't. But I can't quite rule out the possibility that there is some edge case (that needs a test) where it actually matters.


<!-- Sent from Reviewable.io -->

tombogle reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sillsdev/glyssen/763)* status: 29 of 33 files reviewed, 8 unresolved discussions (waiting on @andrew-polk and @tombogle)

___
*[`GlyssenEngine/Character/CharacterUsageStore.cs` line 14 at r1](https://reviewable.io/reviews/sillsdev/glyssen/763#-N5vtyL7AOd2b1SxhNmU:-N5wfL1I7JT2UoWR7n92:b-896fix) ([raw file](https://github.com/sillsdev/glyssen/blob/dfd9f63cca8aa193ff73601f3d08f85efa399be7/GlyssenEngine/Character/CharacterUsageStore.cs#L14)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

This class would benefit from a class comment.

I'm unsure about the name. But I definitely don't have anything better at this point. Maybe the class comment will help...
</blockquote></details>

Done.
___
*[`GlyssenEngine/Character/CharacterUsageStore.cs` line 30 at r1](https://reviewable.io/reviews/sillsdev/glyssen/763#-N5vwKgDCklM_t8zdZ2S:-N5zGMuNC4j2eVZkoieK:b-896fix) ([raw file](https://github.com/sillsdev/glyssen/blob/dfd9f63cca8aa193ff73601f3d08f85efa399be7/GlyssenEngine/Character/CharacterUsageStore.cs#L30)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

This class could also probably benefit from a comment.
</blockquote></details>

Done.
___
*[`GlyssenEngine/Character/CharacterUsageStore.cs` line 90 at r1](https://reviewable.io/reviews/sillsdev/glyssen/763#-N5vvrBZ31lOo68MCaY0:-N5zGXO22qJdo61VnWyf:b-896fix) ([raw file](https://github.com/sillsdev/glyssen/blob/dfd9f63cca8aa193ff73601f3d08f85efa399be7/GlyssenEngine/Character/CharacterUsageStore.cs#L90)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

you left me hanging...

Actually looks like this second sentence just needs to be deleted. You have it again (and complete) later.
</blockquote></details>

Done.
___
*[`GlyssenEngineTests/Character/CharacterUsageStoreTests.cs` line 224 at r1](https://reviewable.io/reviews/sillsdev/glyssen/763#-N5w-Ma72N69BLf3imQq:-N5zw2kTEeagHBuOlF_4:b-896fix) ([raw file](https://github.com/sillsdev/glyssen/blob/dfd9f63cca8aa193ff73601f3d08f85efa399be7/GlyssenEngineTests/Character/CharacterUsageStoreTests.cs#L224)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

My pea-brain needs help on this one. In what sense are the first 3 cases "OnlyMatchingSubstring"?
Those are substrings of an alias perhaps?

Generally speaking for all these tests... is there something which can be done to verify we have the characterID/alias we expect in the test control file?
</blockquote></details>

Done.
___
*[`GlyssenShared/BlockElement.cs` line 98 at r1](https://reviewable.io/reviews/sillsdev/glyssen/763#-N5vlm1vCL3Ade-eiYaR:-N6-20Ju32E0wtM7uIBJ:b-896fix) ([raw file](https://github.com/sillsdev/glyssen/blob/dfd9f63cca8aa193ff73601f3d08f85efa399be7/GlyssenShared/BlockElement.cs#L98)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Don't you mean the same as ContainsNoWords?
</blockquote></details>

Done.



_Code quote:_
```Smalltalk
In almost all situations, this is the opposite of ContainsNoWords, except perhaps for
```


<!-- Sent from Reviewable.io -->

andrew-polk reviewed (changes requested):
> Reviewed 7 of 7 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/glyssen/763)* status: 29 of 33 files reviewed, 1 unresolved discussion (waiting on @tombogle)

___
*[`GlyssenEngine/Character/CharacterUsageStore.cs` line 15 at r2](https://reviewable.io/reviews/sillsdev/glyssen/763#-N6EMJq_5lnlVZnBnpuL:-N6EMJq_5lnlVZnBnpuM:bqt5uia) ([raw file](https://github.com/sillsdev/glyssen/blob/2620dd8f64771e906506eedfcc76ed25594549d4/GlyssenEngine/Character/CharacterUsageStore.cs#L15)):*
> ```Smalltalk
> {
> 	/// <summary>
> 	/// A repository of information about "standard" (as defined by Glyssen) characters and their
> ```

Don't we use the term "standard" elsewhere to mean narrator plus extra-biblical? 
For example, see `CharacterVerseData.StandardCharacter`.

But here, don't you mean anything Glyssen expects based on the control file and narrator/extra-biblical?

This potential ambiguity is also present in the name and comments for the method `GetStandardCharacterName`.

In these new places, how about control, official, expected, normal, regular? (thought of canonical, but that obviously has other ambiguities...)



_Code quote:_
```Smalltalk
"standard"
```


<!-- Sent from Reviewable.io -->

tombogle reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sillsdev/glyssen/763)* status: 27 of 33 files reviewed, 1 unresolved discussion (waiting on @andrew-polk and @tombogle)

___
*[`GlyssenEngine/Character/CharacterUsageStore.cs` line 15 at r2](https://reviewable.io/reviews/sillsdev/glyssen/763#-N6EMJq_5lnlVZnBnpuL:-N6JV-vU4SSWkehmdYI2:b-896fix) ([raw file](https://github.com/sillsdev/glyssen/blob/2620dd8f64771e906506eedfcc76ed25594549d4/GlyssenEngine/Character/CharacterUsageStore.cs#L15)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Don't we use the term "standard" elsewhere to mean narrator plus extra-biblical? 
For example, see `CharacterVerseData.StandardCharacter`.

But here, don't you mean anything Glyssen expects based on the control file and narrator/extra-biblical?

This potential ambiguity is also present in the name and comments for the method `GetStandardCharacterName`.

In these new places, how about control, official, expected, normal, regular? (thought of canonical, but that obviously has other ambiguities...)
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

andrew-polk reviewed (changes requested):
> Reviewed 2 of 33 files at r1.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/glyssen/763)* status: 29 of 33 files reviewed, 9 unresolved discussions (waiting on @andrew-polk and @tombogle)

___
*[`GlyssenEngine/UsxParser.cs` line 155 at r2](https://reviewable.io/reviews/sillsdev/glyssen/763#-N6JY0-6C5LIikH4aadJ:-N6JY0-6C5LIikH4aadK:b-xp537j) ([raw file](https://github.com/sillsdev/glyssen/blob/2620dd8f64771e906506eedfcc76ed25594549d4/GlyssenEngine/UsxParser.cs#L155)):*
> ```Smalltalk
> 		}
> 
> 		private class ExplicitInterruptionInfo : ExplicitQuoteInfo
> ```

Future selves would probably appreciate class comments on these 2 private classes.
___
*[`GlyssenEngine/UsxParser.cs` line 283 at r2](https://reviewable.io/reviews/sillsdev/glyssen/763#-N6JYm5wCAEViKLIgoUe:-N6JYm5wCAEViKLIgoUf:b-jixbib) ([raw file](https://github.com/sillsdev/glyssen/blob/2620dd8f64771e906506eedfcc76ed25594549d4/GlyssenEngine/UsxParser.cs#L283)):*
> ```Smalltalk
> 							ProcessParaChildNode(childNode, sb, ref block, blocks, usxPara.StyleTag);
> 
> 						FlushStringBuilderToBlockElement(sb, block, !(block.BlockElements.LastOrDefault() is QuoteId));
> ```

I find it very helpful for booleans to have labels at the caller.
e.g.
`FlushStringBuilderToBlockElement(sb, block, !(block.BlockElements.LastOrDefault() is QuoteId));`
becomes
`FlushStringBuilderToBlockElement(sb, block, isSomething: !(block.BlockElements.LastOrDefault() is QuoteId));`
___
*[`GlyssenEngine/UsxParser.cs` line 733 at r2](https://reviewable.io/reviews/sillsdev/glyssen/763#-N6J_zBG90nPJMEu5nEz:-N6J_zBG90nPJMEu5nF-:b-e17g2j) ([raw file](https://github.com/sillsdev/glyssen/blob/2620dd8f64771e906506eedfcc76ed25594549d4/GlyssenEngine/UsxParser.cs#L733)):*
> ```Smalltalk
> 					block.CharacterId = m_currentExplicitQuote.StartBlock.CharacterId;
> 					block.CharacterIdInScript = m_currentExplicitQuote.StartBlock.CharacterIdInScript;
> 					// REVIEW: What about the delivery
> ```

This resolved enough for now?



_Code quote:_
```Smalltalk
// REVIEW: What about the delivery
```
___
*[`GlyssenEngineTests/UsxParserTests.cs` line 980 at r2](https://reviewable.io/reviews/sillsdev/glyssen/763#-N6JTY7D5R0YbQlDsZNG:-N6JTY7ECfO5bjxCV0O1:b-ydiwnm) ([raw file](https://github.com/sillsdev/glyssen/blob/2620dd8f64771e906506eedfcc76ed25594549d4/GlyssenEngineTests/UsxParserTests.cs#L980)):*
> ```Smalltalk
> 		[TestCase("John", "Peter")]
> 		[TestCase("John", "Peter (Simon)")]
> 		public void Parse_QtMilestonesForExplicitMemberOfGroup_CharacterInUseSetAsSpecified(
> ```

For tests like this which rely so heavily on the control having a certain value(s), I think it would be helpful to verify the control has that and/or comment explicitly what it has. Otherwise it is somewhat confusing what is actually being tested.
___
*[`GlyssenEngineTests/UsxParserTests.cs` line 2610 at r2](https://reviewable.io/reviews/sillsdev/glyssen/763#-N6JWuiNB4vHhOs0tfEW:-N6JWuiNB4vHhOs0tfEX:b-x86gqn) ([raw file](https://github.com/sillsdev/glyssen/blob/2620dd8f64771e906506eedfcc76ed25594549d4/GlyssenEngineTests/UsxParserTests.cs#L2610)):*
> ```Smalltalk
> 				"should we assume that it is a mistake and ignore it, or should we handle it, as " +
> 				"an opener (perhaps only in the case where the paragraph does not end with " +
> 				"sentence-ending punctuation)?");
> ```

Is this how you want to leave this?
___
*[`GlyssenEngineTests/UsxParserTests.cs` line 2615 at r2](https://reviewable.io/reviews/sillsdev/glyssen/763#-N6JWhh28OHZXs6P3_C8:-N6JWhh28OHZXs6P3_C9:bz2kpij) ([raw file](https://github.com/sillsdev/glyssen/blob/2620dd8f64771e906506eedfcc76ed25594549d4/GlyssenEngineTests/UsxParserTests.cs#L2615)):*
> ```Smalltalk
> 		[TestCase(true)]
> 		[TestCase(false)]
> 		public void Parse_QtMilestonesLeftOpenFollowedByHebrewSubtitle_WjClosesOpenMilestoneQuote(bool includeSectionHead)
> ```

I don't understand this test. Is it just mislabeled? (There is no wj)
___
*[`GlyssenEngineTests/UsxParserTests.cs` line 2729 at r2](https://reviewable.io/reviews/sillsdev/glyssen/763#-N6JWywdBnFnQQAroQwI:-N6JWywdBnFnQQAroQwJ:b-x86gqn) ([raw file](https://github.com/sillsdev/glyssen/blob/2620dd8f64771e906506eedfcc76ed25594549d4/GlyssenEngineTests/UsxParserTests.cs#L2729)):*
> ```Smalltalk
> 		public void Parse_QtMilestonesQuoteOpenPastLastVerseWhereCharacterIsExpected_LastBlockForExpectedVerseAndBlockPrecedingCloseNeedsReview()
> 		{
> 			Assert.Ignore("Write this test (unless we determine that the QuoteParser can handle this case.");
> ```

Is this how you want to leave this?
___
*[`GlyssenEngineTests/UsxParserTests.cs` line 2735 at r2](https://reviewable.io/reviews/sillsdev/glyssen/763#-N6JWzy0AqUQvdHN7idH:-N6JWzy0AqUQvdHN7idI:b-x86gqn) ([raw file](https://github.com/sillsdev/glyssen/blob/2620dd8f64771e906506eedfcc76ed25594549d4/GlyssenEngineTests/UsxParserTests.cs#L2735)):*
> ```Smalltalk
> 		public void Parse_QtMilestonesNotClosed_LastBlockForExpectedVerseNeedsReview()
> 		{
> 			Assert.Ignore("Write this test (unless we determine that the QuoteParser can handle this case.");
> ```

Is this how you want to leave this?


<!-- Sent from Reviewable.io -->

andrew-polk reviewed (changes requested):
First pass completed.

> Reviewed 2 of 33 files at r1, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/glyssen/763#-:-N6JhtKa2wVbs4gFwxni:b-nxgybn)* status: all files reviewed, 13 unresolved discussions (waiting on @tombogle)

___
*[`GlyssenEngine/Character/CharacterUsageStore.cs` line 15 at r2](https://reviewable.io/reviews/sillsdev/glyssen/763#-N6EMJq_5lnlVZnBnpuL:-N6Jb13N4lcTM2Bl9v3C:b-r34jur) ([raw file](https://github.com/sillsdev/glyssen/blob/2620dd8f64771e906506eedfcc76ed25594549d4/GlyssenEngine/Character/CharacterUsageStore.cs#L15)):*
<details><summary><i>Previously, tombogle (Tom Bogle) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Shouldn't `GetStandardCharacterName` become `GetOfficialCharacterName`?



_Code quote:_
```Smalltalk
GetStandardCharacterName
```
___
*[`GlyssenEngine/Quote/QuoteParser.cs` line 612 at r3](https://reviewable.io/reviews/sillsdev/glyssen/763#-N6JhMFP2ytjjgOjnYWB:-N6JhMFP2ytjjgOjnYWC:b-bfjh74) ([raw file](https://github.com/sillsdev/glyssen/blob/ee000db6e8d271f3b77b28c711b9094e7a9e33c1/GlyssenEngine/Quote/QuoteParser.cs#L612)):*
> ```Smalltalk
> 			}
> 
> 			// TODO: Write unit test with special character style following explicit quote following parser-detected quote
> ```

to-done?
___
*[`GlyssenEngine/Quote/QuoteParser.cs` line 1026 at r3](https://reviewable.io/reviews/sillsdev/glyssen/763#-N6JhgzW-ZYs-OMaq3iT:-N6JhgzW-ZYs-OMaq3iU:babwjh8) ([raw file](https://github.com/sillsdev/glyssen/blob/ee000db6e8d271f3b77b28c711b9094e7a9e33c1/GlyssenEngine/Quote/QuoteParser.cs#L1026)):*
> ```Smalltalk
> 		/// and flush the current block elements to a block
> 		/// </summary>
> 		private void FlushStringBuilderAndBlock(StringBuilder sb, ref string styleTag, bool nonNarrator, bool characterUnknown = false)
> ```

With `styleTag` being made a `ref` parameter, I was expecting to see it potentially modified.
I don't see that.
___
*[`GlyssenEngineTests/Quote/QuoteParserTests.cs` line 320 at r3](https://reviewable.io/reviews/sillsdev/glyssen/763#-N6JcoUA237Ab9_4tvjV:-N6JcoUA237Ab9_4tvjW:bm9uaxc) ([raw file](https://github.com/sillsdev/glyssen/blob/ee000db6e8d271f3b77b28c711b9094e7a9e33c1/GlyssenEngineTests/Quote/QuoteParserTests.cs#L320)):*
> ```Smalltalk
> 			Assert.IsFalse(output[i].IsPredeterminedFirstLevelQuoteEnd);
> 			Assert.AreEqual(MultiBlockQuote.None, output[i].MultiBlockQuote);
> 			Assert.AreEqual(input[i + 2].GetText(true, true), output[i].GetText(true, true));
> ```

I always find these repetitive list of asserts pretty hard to follow. Do you think it would be helpful/worth it to do something like
```
VerifyResultBlock(
  block,
  isPredeterminedStart: true,
  character: Jesus,
  multiBlock: MultiBlockQuote.Continuation  
)
```
etc?
Maybe there isn't enough consistency between what needs to be verified to make it worth it.
___
*[`GlyssenEngineTests/Quote/QuoteParserTests.cs` line 7298 at r3](https://reviewable.io/reviews/sillsdev/glyssen/763#-N6JfPdbCmPYvrmMMvfd:-N6JfPdbCmPYvrmMMvfe:b-4i0opd) ([raw file](https://github.com/sillsdev/glyssen/blob/ee000db6e8d271f3b77b28c711b9094e7a9e33c1/GlyssenEngineTests/Quote/QuoteParserTests.cs#L7298)):*
> ```Smalltalk
> 		#endregion
> 
> 		#region PG-1419 tests (blocks created by quote milestones)
> ```

Is it worth noting in each place that there are other milestone tests in the other fixture and why we need these or those in the particular fixture?
Feel free to say no.


<!-- Sent from Reviewable.io -->

tombogle reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sillsdev/glyssen/763)* status: all files reviewed, 13 unresolved discussions (waiting on @andrew-polk)

___
*[`GlyssenEngine/UsxParser.cs` line 155 at r2](https://reviewable.io/reviews/sillsdev/glyssen/763#-N6JY0-6C5LIikH4aadJ:-N6KgBRP10I-gTIws1B2:b-896fix) ([raw file](https://github.com/sillsdev/glyssen/blob/2620dd8f64771e906506eedfcc76ed25594549d4/GlyssenEngine/UsxParser.cs#L155)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Future selves would probably appreciate class comments on these 2 private classes.
</blockquote></details>

Done.
___
*[`GlyssenEngine/UsxParser.cs` line 283 at r2](https://reviewable.io/reviews/sillsdev/glyssen/763#-N6JYm5wCAEViKLIgoUe:-N6Kh3Kk-kNBjttMKKXa:b-896fix) ([raw file](https://github.com/sillsdev/glyssen/blob/2620dd8f64771e906506eedfcc76ed25594549d4/GlyssenEngine/UsxParser.cs#L283)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

I find it very helpful for booleans to have labels at the caller.
e.g.
`FlushStringBuilderToBlockElement(sb, block, !(block.BlockElements.LastOrDefault() is QuoteId));`
becomes
`FlushStringBuilderToBlockElement(sb, block, isSomething: !(block.BlockElements.LastOrDefault() is QuoteId));`
</blockquote></details>

Done.
___
*[`GlyssenEngine/UsxParser.cs` line 733 at r2](https://reviewable.io/reviews/sillsdev/glyssen/763#-N6J_zBG90nPJMEu5nEz:-N6KhXKx3mW2_Ennp8Uy:b-lk7jc5) ([raw file](https://github.com/sillsdev/glyssen/blob/2620dd8f64771e906506eedfcc76ed25594549d4/GlyssenEngine/UsxParser.cs#L733)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

This resolved enough for now?
</blockquote></details>

Expanded the comment to make the question more explicit.
___
*[`GlyssenEngine/Character/CharacterUsageStore.cs` line 15 at r2](https://reviewable.io/reviews/sillsdev/glyssen/763#-N6EMJq_5lnlVZnBnpuL:-N6Khv3Z0GxBkFa1IOzX:bsyo01z) ([raw file](https://github.com/sillsdev/glyssen/blob/2620dd8f64771e906506eedfcc76ed25594549d4/GlyssenEngine/Character/CharacterUsageStore.cs#L15)):*
> GetStandardCharacterName
Or GetKnownCharacterName? Or GetRecognizedCharacterName?
___
*[`GlyssenEngine/Quote/QuoteParser.cs` line 612 at r3](https://reviewable.io/reviews/sillsdev/glyssen/763#-N6JhMFP2ytjjgOjnYWB:-N6OBuCC4IZpc_tWLTWQ:bwhu7y) ([raw file](https://github.com/sillsdev/glyssen/blob/ee000db6e8d271f3b77b28c711b9094e7a9e33c1/GlyssenEngine/Quote/QuoteParser.cs#L612)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

to-done?
</blockquote></details>

Now it is.



_Code quote:_
```Smalltalk
TODO: Write unit test with special character
```
___
*[`GlyssenEngine/Quote/QuoteParser.cs` line 1026 at r3](https://reviewable.io/reviews/sillsdev/glyssen/763#-N6JhgzW-ZYs-OMaq3iT:-N6OBwQm9NjCyRo_OjF1:b-lqiuj1) ([raw file](https://github.com/sillsdev/glyssen/blob/ee000db6e8d271f3b77b28c711b9094e7a9e33c1/GlyssenEngine/Quote/QuoteParser.cs#L1026)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

With `styleTag` being made a `ref` parameter, I was expecting to see it potentially modified.
I don't see that.
</blockquote></details>

Looks like I back-tracked on that.
___
*[`GlyssenEngineTests/UsxParserTests.cs` line 980 at r2](https://reviewable.io/reviews/sillsdev/glyssen/763#-N6JTY7D5R0YbQlDsZNG:-N6OHwDpCW-0W6XR0Fu0:b-896fix) ([raw file](https://github.com/sillsdev/glyssen/blob/2620dd8f64771e906506eedfcc76ed25594549d4/GlyssenEngineTests/UsxParserTests.cs#L980)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

For tests like this which rely so heavily on the control having a certain value(s), I think it would be helpful to verify the control has that and/or comment explicitly what it has. Otherwise it is somewhat confusing what is actually being tested.
</blockquote></details>

Done.



_Code quote:_
```Smalltalk
arse_QtMilestonesForExplicitMemberOfGroup_CharacterInUseSetAsSpecified
```
___
*[`GlyssenEngineTests/UsxParserTests.cs` line 2610 at r2](https://reviewable.io/reviews/sillsdev/glyssen/763#-N6JWuiNB4vHhOs0tfEW:-N6OIg2tEMD5LdItoRnF:b-3ftto8) ([raw file](https://github.com/sillsdev/glyssen/blob/2620dd8f64771e906506eedfcc76ed25594549d4/GlyssenEngineTests/UsxParserTests.cs#L2610)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Is this how you want to leave this?
</blockquote></details>

Added more information.



_Code quote:_
```Smalltalk
ssert.Inconclusive("REVIEW: if we encounter
```
___
*[`GlyssenEngineTests/UsxParserTests.cs` line 2615 at r2](https://reviewable.io/reviews/sillsdev/glyssen/763#-N6JWhh28OHZXs6P3_C8:-N6OJYKo5S6brpNbY82U:b-9h8srt) ([raw file](https://github.com/sillsdev/glyssen/blob/2620dd8f64771e906506eedfcc76ed25594549d4/GlyssenEngineTests/UsxParserTests.cs#L2615)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

I don't understand this test. Is it just mislabeled? (There is no wj)
</blockquote></details>

Fixed name of test.
___
*[`GlyssenEngineTests/UsxParserTests.cs` line 2729 at r2](https://reviewable.io/reviews/sillsdev/glyssen/763#-N6JWywdBnFnQQAroQwI:-N6OLPsm2jzjCv1zXvXA:b-h0w7tz) ([raw file](https://github.com/sillsdev/glyssen/blob/2620dd8f64771e906506eedfcc76ed25594549d4/GlyssenEngineTests/UsxParserTests.cs#L2729)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Is this how you want to leave this?
</blockquote></details>

Looks like the needed test is in place in QuoteParserTests
___
*[`GlyssenEngineTests/UsxParserTests.cs` line 2735 at r2](https://reviewable.io/reviews/sillsdev/glyssen/763#-N6JWzy0AqUQvdHN7idH:-N6OLcY80T3NpUqf2X20:b-ns4fnv) ([raw file](https://github.com/sillsdev/glyssen/blob/2620dd8f64771e906506eedfcc76ed25594549d4/GlyssenEngineTests/UsxParserTests.cs#L2735)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Is this how you want to leave this?
</blockquote></details>

Looks like the needed test is in place in QuoteParserTests.
___
*[`GlyssenEngineTests/Quote/QuoteParserTests.cs` line 320 at r3](https://reviewable.io/reviews/sillsdev/glyssen/763#-N6JcoUA237Ab9_4tvjV:-N6OLqiM1FQ1ri3U3Y5F:bgc0oz1) ([raw file](https://github.com/sillsdev/glyssen/blob/ee000db6e8d271f3b77b28c711b9094e7a9e33c1/GlyssenEngineTests/Quote/QuoteParserTests.cs#L320)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

I always find these repetitive list of asserts pretty hard to follow. Do you think it would be helpful/worth it to do something like
```
VerifyResultBlock(
  block,
  isPredeterminedStart: true,
  character: Jesus,
  multiBlock: MultiBlockQuote.Continuation  
)
```
etc?
Maybe there isn't enough consistency between what needs to be verified to make it worth it.
</blockquote></details>

In general, we tend to check different things. We could check things more consistently and come up with reasonable defaults for the less important stuff or try to write multiple semi-meaningful methods for checking different sets of things. For now, I'll at least try to separate the lines into groups of checks with blank lines more consistently to make it easier to follow.
___
*[`GlyssenEngineTests/Quote/QuoteParserTests.cs` line 7298 at r3](https://reviewable.io/reviews/sillsdev/glyssen/763#-N6JfPdbCmPYvrmMMvfd:-N6OtCOY4Oqz7Y1l8rqo:bte3nvz) ([raw file](https://github.com/sillsdev/glyssen/blob/ee000db6e8d271f3b77b28c711b9094e7a9e33c1/GlyssenEngineTests/Quote/QuoteParserTests.cs#L7298)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Is it worth noting in each place that there are other milestone tests in the other fixture and why we need these or those in the particular fixture?
Feel free to say no.
</blockquote></details>

I did it for these. I was surprised, but I didn't find any other regions that had a group of tests in each of the two fixtures.



_Code quote:_
```Smalltalk
#region PG-1419 tests (blocks created by quote milestones)
```


<!-- Sent from Reviewable.io -->

andrew-polk reviewed (changes requested):
Looks good. Just a few nits. Feel free to merge after the method name change.

> Reviewed 5 of 5 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/glyssen/763#-:-N6P0XlvD8EFAtxGwg-e:bqbm821)* status: all files reviewed, 4 unresolved discussions (waiting on @tombogle)

___
*[`GlyssenEngine/UsxParser.cs` line 185 at r4](https://reviewable.io/reviews/sillsdev/glyssen/763#-N6OzvrvARkWs7eVQkIh:-N6OzvrvARkWs7eVQkIi:bke26wo) ([raw file](https://github.com/sillsdev/glyssen/blob/099b37b58bc8989be9b95eb64eda1641b765aca8/GlyssenEngine/UsxParser.cs#L185)):*
> ```Smalltalk
> 		/// <summary>
> 		/// A class for the special-case where the "quote" is actually an
> 		/// interruption (by the narrator) of an quote.
> ```

a quote (if making changes)



_Code quote:_
```Smalltalk
an quote
```
___
*[`GlyssenEngine/Character/CharacterUsageStore.cs` line 15 at r2](https://reviewable.io/reviews/sillsdev/glyssen/763#-N6EMJq_5lnlVZnBnpuL:-N6P-3RO6inIikKUVpoO:btqhi46) ([raw file](https://github.com/sillsdev/glyssen/blob/2620dd8f64771e906506eedfcc76ed25594549d4/GlyssenEngine/Character/CharacterUsageStore.cs#L15)):*
<details><summary><i>Previously, tombogle (Tom Bogle) wrote…</i></summary><blockquote>

> GetStandardCharacterName
Or GetKnownCharacterName? Or GetRecognizedCharacterName?
</blockquote></details>

Any are fine. Known may be best.
___
*[`GlyssenEngineTests/UsxParserTests.cs` line 934 at r4](https://reviewable.io/reviews/sillsdev/glyssen/763#-N6P-A5T5ka1DotRVSIC:-N6P-A5T5ka1DotRVSID:b9i7714) ([raw file](https://github.com/sillsdev/glyssen/blob/099b37b58bc8989be9b95eb64eda1641b765aca8/GlyssenEngineTests/UsxParserTests.cs#L934)):*
> ```Smalltalk
> 		}
> 
> 		bool HasAPeterAndHJohnCharacter(ICharacterDeliveryInfo cv)
> ```

HasAPeterAndJohnCharacter?



_Suggestion:_
```Smalltalk
HasAPeterAndHJohnCharacter
```
___
*[`GlyssenEngineTests/UsxParserTests.cs` line 2659 at r4](https://reviewable.io/reviews/sillsdev/glyssen/763#-N6OzputDtaOoZigSCcj:-N6OzputDtaOoZigSCck:b-3uo4ve) ([raw file](https://github.com/sillsdev/glyssen/blob/099b37b58bc8989be9b95eb64eda1641b765aca8/GlyssenEngineTests/UsxParserTests.cs#L2659)):*
> ```Smalltalk
> 				"handle it as an opener (perhaps only in the case where the paragraph does not " +
> 				"end with sentence-ending punctuation)? We should probably wait to resolve this " +
> 				"question until we actually encounter is on some real data and see what the " +
> ```

it (if you're making more changes)



_Code quote:_
```Smalltalk
is
```


<!-- Sent from Reviewable.io -->

andrew-polk reviewed (approved):
> Reviewed 5 of 5 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/glyssen/763)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @tombogle)



<!-- Sent from Reviewable.io -->
